src/tests/ascent/t_ascent_expressions.cpp

```diff
@@ -44,628 +44,939 @@
+  Node multi_dom;
+  blueprint::mesh::to_multi_domain(data, multi_dom);
+
+  runtime::expressions::register_builtin();
+  runtime::expressions::ExpressionEval eval(&multi_dom);
+
+  std::string expr;
+
+  string output_path = prepare_output_dir();
+
+  expr = "binning('braid', 'sum', [axis('x'), axis('y')], output='bins')";
+  eval.evaluate(expr);
+  expr = "binning('braid', 'std', [axis('x', num_bins=10), axis('y', num_bins=10)], output='mesh')";
+  eval.evaluate(expr);
+
+  std::string output_file =
+      conduit::utils::join_file_path(output_path, "tout_binning_braid_xysum");
+  output_pseudocolor(multi_dom, "braid_sum", output_file);
+  EXPECT_TRUE(check_test_image(output_file, 0.01));
+
+  output_file = conduit::utils::join_file_path(
+      output_path, "tout_binning_painted_braid_xystd");
+  output_pseudocolor(multi_dom, "painted_braid_std", output_file);
+  EXPECT_TRUE(check_test_image(output_file, 0.01));
 }
 
-//-----------------------------------------------------------------------------
-TEST(ascent_expressions, if_expressions)
+TEST(ascent_binning, multi_dom_binning)
 {
-    Node n;
-    ascent::about(n);
-
-    //
-    // Create an example mesh.
-    //
-    Node data, verify_info;
-    conduit::blueprint::mesh::examples::braid("hexs",
-                                              EXAMPLE_MESH_SIDE_DIM,
-                                              EXAMPLE_MESH_SIDE_DIM,
-                                              EXAMPLE_MESH_SIDE_DIM,
-                                              data);
-    // ascent normally adds this but we are doing an end around
-    data["state/domain_id"] = 0;
-    Node multi_dom;
-    blueprint::mesh::to_multi_domain(data, multi_dom);
-
-    runtime::expressions::register_builtin();
-    runtime::expressions::ExpressionEval eval(&multi_dom);
-
-    conduit::Node res;
-    std::string expr;
-
-    expr = "if 5 == 5 then 1 else 2";
-    res = eval.evaluate(expr);
-    EXPECT_EQ(res["value"].to_int32(), 1);
-    EXPECT_EQ(res["type"].as_string(), "int");
+  // the vtkm runtime is currently our only rendering runtime
+  Node n;
+  ascent::about(n);
+  // only run this test if ascent was built with vtkm support
+  if(n["runtimes/ascent/vtkm/status"].as_string() == "disabled")
+  {
+    ASCENT_INFO("Ascent support disabled, skipping test");
+    return;
+  }
+
+  //
+  // Create an example mesh.
+  //
+  Node data, verify_info;
+  conduit::blueprint::mesh::examples::spiral(5, data);
+  Node multi_dom;
+  blueprint::mesh::to_multi_domain(data, multi_dom);
+  // ascent normally adds this but we are doing an end around
+  for(int i = 0; i < multi_dom.number_of_children(); ++i)
+  {
+    multi_dom.child(i)["state/cycle"] = 100;
+    multi_dom.child(i)["state/domain_id"] = 0;
+  }
+
+  runtime::expressions::register_builtin();
+  runtime::expressions::ExpressionEval eval(&multi_dom);
+
+  std::string expr;
+  expr = "binning('dist', 'std', [axis('x'), axis('y')], output='bins')";
```

Comment:
these cases use default # of bins?

Reply:
If the number of bins is not specified it will try to calculate a reasonable number of bins. If the topo is uniform and the axis is x/y/z that comes from dim. For non-uniform meshes and x/y/z the number of bins is max_x - min_x. For fields it's just 256. Maybe it's better to just have the default be 256 everywhere in case it gets too big.

Min and max are also calculated if they aren't specified. For x/y/z it looks at all the fields and calculates the bounds from that.

Reply:
auto min and max is good, 

for the # of bins, a default # is easier to reason about. 

We don't want an unstructured astro code modeling the universe to allocate 1e+20 bins :-)